### PR TITLE
flux-jobs: add --recursive option

### PR DESCRIPTION
**Pull request by grondo**
This PR adds a very simple `-R, --recursive` option to `flux-jobs` which lists child jobs recursively.

Similar to `ls -R`, when listing recursive jobs, each child instance of Flux is prefixed with its "path", which also happens to be a suitable jobid URI for use with `flux-uri` (and thus `flux-proxy(1)` and `flux-top(1)`).

If `--stats` is given, then the jobid path is also accompanied by that job's statistics.

A `-L, --level=N` option can be used to stop the recursion at a defined depth.

As part of this work, a convenience `uri` property was introduced for `JobInfo` objects, along with support for listing `{uri}` and `{uri.local}` attributes in `flux-jobs`.

Finally, in keeping with the visual cue introduced recently in `flux-top(1)` running jobs which are also instances of Flux are now colored blue in `flux-jobs(1)` output.

Example:

```console
$ flux jobs
       JOBID USER     NAME       ST NTASKS NNODES  RUNTIME NODELIST
  ƒD4Z3ADV9H grondo   flux        R      2      1   6.974m asp

$ flux jobs -Rf running
       JOBID USER     NAME       ST NTASKS NNODES  RUNTIME NODELIST
  ƒD4Z3ADV9H grondo   flux        R      2      1    7.08m asp

ƒD4Z3ADV9H:
     ƒdRdMaQ grondo   flux        R      1      1   7.044m asp
     ƒdRdMaP grondo   flux        R      1      1   7.044m asp

ƒD4Z3ADV9H/ƒdRdMaQ:
     ƒSWaeF9 grondo   flux        R      1      1   7.016m asp
     ƒSV6exo grondo   flux        R      1      1   7.016m asp

ƒD4Z3ADV9H/ƒdRdMaQ/ƒSWaeF9:
     ƒSZYcoq grondo   flux        R      1      1   6.987m asp

ƒD4Z3ADV9H/ƒdRdMaQ/ƒSWaeF9/ƒSZYcoq:
     ƒT5CNSj grondo   flux        R      1      1   6.958m asp

ƒD4Z3ADV9H/ƒdRdMaQ/ƒSWaeF9/ƒSZYcoq/ƒT5CNSj:
     ƒTQUD7D grondo   flux        R      1      1   6.928m asp

ƒD4Z3ADV9H/ƒdRdMaQ/ƒSWaeF9/ƒSZYcoq/ƒT5CNSj/ƒTQUD7D:
     ƒU3Xu9q grondo   flux        R      1      1   6.899m asp

ƒD4Z3ADV9H/ƒdRdMaQ/ƒSWaeF9/ƒSZYcoq/ƒT5CNSj/ƒTQUD7D/ƒU3Xu9q:
     ƒSyGRKM grondo   flux        R      1      1    6.87m asp

ƒD4Z3ADV9H/ƒdRdMaQ/ƒSWaeF9/ƒSZYcoq/ƒT5CNSj/ƒTQUD7D/ƒU3Xu9q/ƒSyGRKM:
    ƒ2gN1qQT grondo   sleep       R      1      1   0.033s asp

ƒD4Z3ADV9H/ƒdRdMaQ/ƒSV6exo:
     ƒm6Pe7q grondo   flux        R      1      1   6.965m asp

ƒD4Z3ADV9H/ƒdRdMaQ/ƒSV6exo/ƒm6Pe7q:
     ƒnGb55h grondo   flux        R      1      1    6.91m asp

ƒD4Z3ADV9H/ƒdRdMaQ/ƒSV6exo/ƒm6Pe7q/ƒnGb55h:
     ƒnVvxco grondo   flux        R      1      1   6.855m asp

ƒD4Z3ADV9H/ƒdRdMaQ/ƒSV6exo/ƒm6Pe7q/ƒnGb55h/ƒnVvxco:
     ƒoaCSTH grondo   flux        R      1      1   6.798m asp

ƒD4Z3ADV9H/ƒdRdMaQ/ƒSV6exo/ƒm6Pe7q/ƒnGb55h/ƒnVvxco/ƒoaCSTH:
     ƒnw8kQf grondo   flux        R      1      1   6.742m asp

ƒD4Z3ADV9H/ƒdRdMaQ/ƒSV6exo/ƒm6Pe7q/ƒnGb55h/ƒnVvxco/ƒoaCSTH/ƒnw8kQf:
    ƒ3djD1oU grondo   sleep       R      1      1   0.332s asp

ƒD4Z3ADV9H/ƒdRdMaP:
     ƒSjvXnF grondo   flux        R      1      1   7.016m asp

ƒD4Z3ADV9H/ƒdRdMaP/ƒSjvXnF:
     ƒqhFQqu grondo   flux        R      1      1   6.962m asp

ƒD4Z3ADV9H/ƒdRdMaP/ƒSjvXnF/ƒqhFQqu:
     ƒrLK6tX grondo   flux        R      1      1   6.906m asp

ƒD4Z3ADV9H/ƒdRdMaP/ƒSjvXnF/ƒqhFQqu/ƒrLK6tX:
     ƒrSF41u grondo   flux        R      1      1    6.85m asp

ƒD4Z3ADV9H/ƒdRdMaP/ƒSjvXnF/ƒqhFQqu/ƒrLK6tX/ƒrSF41u:
     ƒrLK6tX grondo   flux        R      1      1   6.794m asp

ƒD4Z3ADV9H/ƒdRdMaP/ƒSjvXnF/ƒqhFQqu/ƒrLK6tX/ƒrSF41u/ƒrLK6tX:
     ƒqx5HfM grondo   flux        R      1      1   6.737m asp

ƒD4Z3ADV9H/ƒdRdMaP/ƒSjvXnF/ƒqhFQqu/ƒrLK6tX/ƒrSF41u/ƒrLK6tX/ƒqx5HfM:
    ƒ4hjryTu grondo   sleep       R      1      1   0.380s asp


$ flux jobs -Rf running -L2 --stats
1 running, 10 completed, 6 failed, 0 pending
       JOBID USER     NAME       ST NTASKS NNODES  RUNTIME NODELIST
  ƒD4Z3ADV9H grondo   flux        R      2      1   7.434m asp

ƒD4Z3ADV9H:
2 running, 0 completed, 0 failed, 1 pending
     ƒdRdMaQ grondo   flux        R      1      1   7.398m asp
     ƒdRdMaP grondo   flux        R      1      1   7.398m asp

ƒD4Z3ADV9H/ƒdRdMaQ:
2 running, 0 completed, 0 failed, 0 pending
     ƒSWaeF9 grondo   flux        R      1      1   7.369m asp
     ƒSV6exo grondo   flux        R      1      1   7.369m asp

ƒD4Z3ADV9H/ƒdRdMaP:
1 running, 0 completed, 0 failed, 1 pending
     ƒSjvXnF grondo   flux        R      1      1   7.369m asp
```

### CODE REVIEW

**Comment by grondo** [comment]
BTW, meant to mention that I tried parallelizing the fetch of the job list for each child job at any given level using a `concurrent.futures.ThreadPool`, but I didn't really see much speedup. It is possible that I was doing something wrong, but it would be trivial to add back later if the serialized RPCs become an issue.

File: src/bindings/python/flux/uri/uri.py [hunk 1]
```diff
@@ -8,6 +8,7 @@
 # SPDX-License-Identifier: LGPL-3.0
```
Comment (chu11) [inline src/bindings/python/flux/uri/uri.py:8]: commit message typo "fi" -> "if"

`Then, in the JobURI str() method, return the local URI fi force_local is True.`

File: t/t2800-jobs-recursive.t [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+test_expect_success 'start a recursive job' '
+	id=$(flux mini submit flux start /bin/true) &&
+	rid=$(flux mini submit -n2 \
+		flux start \
+		flux mini submit --wait --cc=1-2 flux start \
+			"flux mini submit sleep inf && \
+                 	 touch ready.\$FLUX_JOB_CC && \
+		 	 flux queue idle") &&
+	flux job wait-event $id clean
+'
```
Comment (chu11) [inline t/t2800-jobs-recursive.t:22]: maybe add a small comment here, takes a second to grok (i think 1 sub instance that is finished, 1 sub instance that has 2 sub instances within it).  Not sure of the purpose of the idle queue?

**Review by chu11 — COMMENTED** [review]
this looks awesome.  took a quick skim and nothing stood out to me other than some nits.

File: t/t2800-jobs-recursive.t (hunk 2 above)
Comment (grondo) [reply t/t2800-jobs-recursive.t:22]: `flux queue idle` waits for the running job to finish. Otherwise instance exits and job is killed.

Comment (grondo) [reply t/t2800-jobs-recursive.t:22]: Good point, I'll add a comment before starting that test workload.

**Comment by grondo** [comment]
Thanks for the quick feedback @chu11! I've addressed the review comments and force-pushed result.

**Review by chu11 — APPROVED** [review]
LGTM!

**Comment by grondo** [comment]
Found and fixed a small bug (recursive option always requires `status` as well as `annotations` attributes).

Will set MWP unless I hear any objections (have another PR ready based on this one)

**Comment by garlick** [comment]
This is great!

**Comment by grondo** [comment]
Thanks!

**Pull request merged**